### DISCUSSION TITLE
Persistent Circle Deformation Issue CNC - Seeking Further Assistance SOS!

### DISCUSSION THREAD

**Discussion by Serhai** [winder/Universal-G-Code-Sender#2426] [Help]
Dear forum members,

I am reaching out again for your assistance. I've built a CNC machine, a Workbee V2 1500*1500 from Bulkman 3D, and I am encountering a persistent issue: my circles are not perfectly round. Here is a summary of my setup and the troubleshooting attempts I've made:

    CNC: Workbee V2 1500*1500 from Bulkman 3D, assembled by myself.
    Controller: Arduino Nano clone (GRBL v1.1).
    Motors: Nema 23HD86001Y-21B.
    Stepper Motor Drivers: TB6600.

Despite various adjustments and advice, the problem remains:

    Adjustments to size settings, speed, scale, and motor microsteps.
    Optimization of the machine's linear movements.
    Replacement of the controller software (USG and Openbuild) and use of Inkscape and Carbide for design.
    GRBL adjustments (setting $0, modifying $11 and $12, precise adjustments of $100, $101, and $102).
    Modification of the drivers' microsteps from 4 to 8, with recalibration for millimeter precision.

Unfortunately, the machine still shows a recoil movement and deforms the circles, even though the G-code seems correct. Do you have any additional ideas or specific aspects I could further explore?

 My setting:
>>> $$
$0 = 3 (Step pulse time, microseconds)
$1 = 255 (Step idle delay, milliseconds)
$2 = 0 (Step pulse invert, mask)
$3 = 3 (Step direction invert, mask)
$4 = 0 (Invert step enable pin, boolean)
$5 = 0 (Invert limit pins, boolean)
$6 = 0 (Invert probe pin, boolean)
$10 = 2 (Status report options, mask)
$11 = 0.010 (Junction deviation, millimeters)
$12 = 0.005 (Arc tolerance, millimeters)
$13 = 0 (Report in inches, boolean)
$20 = 0 (Soft limits enable, boolean)
$21 = 1 (Hard limits enable, boolean)
$22 = 1 (Homing cycle enable, boolean)
$23 = 7 (Homing direction invert, mask)
$24 = 180.000 (Homing locate feed rate, mm/min)
$25 = 1800.000 (Homing search seek rate, mm/min)
$26 = 250 (Homing switch debounce delay, milliseconds)
$27 = 2.000 (Homing switch pull-off distance, millimeters)
$30 = 0 (Maximum spindle speed, RPM)
$31 = 0 (Minimum spindle speed, RPM)
$32 = 0 (Laser-mode enable, boolean)
$100 = 200.700 (X-axis travel resolution, step/mm)
$101 = 200.300 (Y-axis travel resolution, step/mm)
$102 = 100.000 (Z-axis travel resolution, step/mm)
$110 = 2000.000 (X-axis maximum rate, mm/min)
$111 = 2000.000 (Y-axis maximum rate, mm/min)
$112 = 1000.000 (Z-axis maximum rate, mm/min)
$120 = 150.000 (X-axis acceleration, mm/sec^2)
$121 = 150.000 (Y-axis acceleration, mm/sec^2)
$122 = 150.000 (Z-axis acceleration, mm/sec^2)
$130 = 1300.000 (X-axis maximum travel, millimeters)
$131 = 1200.000 (Y-axis maximum travel, millimeters)
$132 = 48.000 (Z-axis maximum travel, millimeters)

and the Gcode of circle:
G90;
G21;
Z30
G1 F1500.0 X324.217290 Y157.909780;
Z0
G1 F600.0 X324.213629 Y157.494210;
G1 X324.202648 Y157.078769;
G1 X324.184349 Y156.663586;
G1 X324.158739 Y156.248790;
G1 X324.125825 Y155.834510;
G1 X324.085618 Y155.420874;
G1 X324.038130 Y155.008010;
G1 X323.983376 Y154.596046;
G1 X323.921373 Y154.185112;
G1 X323.852140 Y153.775333;
G1 X323.775699 Y153.366838;
G1 X323.692073 Y152.959753;
G1 X323.601288 Y152.554204;
G1 X323.503373 Y152.150318;
G1 X323.398357 Y151.748219;
G1 X323.286274 Y151.348033;
G1 X323.167159 Y150.949883;
G1 X323.041047 Y150.553894;
G1 X322.907979 Y150.160187;
G1 X322.767996 Y149.768887;
G1 X322.621141 Y149.380113;
G1 X322.467459 Y148.993986;
G1 X322.306999 Y148.610627;
G1 X322.139810 Y148.230154;
G1 X321.965944 Y147.852686;
G1 X321.785455 Y147.478339;
G1 X321.598400 Y147.107230;
G1 X321.404835 Y146.739474;
G1 X321.204822 Y146.375186;
G1 X320.998422 Y146.014477;
G1 X320.785699 Y145.657461;
G1 X320.566720 Y145.304247;
G1 X320.341552 Y144.954946;
G1 X320.110266 Y144.609666;
G1 X319.872932 Y144.268515;
G1 X319.629625 Y143.931597;
G1 X319.380421 Y143.599018;
G1 X319.125396 Y143.270881;
G1 X318.864629 Y142.947288;
G1 X318.598203 Y142.628339;
G1 X318.326199 Y142.314133;
G1 X318.048701 Y142.004768;
G1 X317.765797 Y141.700339;
G1 X317.477573 Y141.400942;
G1 X317.184119 Y141.106669;
G1 X316.885527 Y140.817611;
G1 X316.581889 Y140.533859;
G1 X316.273299 Y140.255499;
G1 X315.959853 Y139.982620;
G1 X315.641648 Y139.715305;
G1 X315.318783 Y139.453637;
G1 X314.991359 Y139.197698;
G1 X314.659476 Y138.947566;
G1 X314.323239 Y138.703321;
G1 X313.982750 Y138.465037;
G1 X313.638117 Y138.232788;
G1 X313.289445 Y138.006647;
G1 X312.936844 Y137.786683;
G1 X312.580422 Y137.572966;
G1 X312.220291 Y137.365560;
G1 X311.856561 Y137.164532;
G1 X311.489347 Y136.969942;
G1 X311.118761 Y136.781852;
G1 X310.744919 Y136.600320;
G1 X310.367937 Y136.425402;
G1 X309.987932 Y136.257152;
G1 X309.605022 Y136.095624;
G1 X309.219326 Y135.940866;
G1 X308.830963 Y135.792927;
G1 X308.440054 Y135.651852;
G1 X308.046721 Y135.517687;
G1 X307.651085 Y135.390471;
G1 X307.253269 Y135.270246;
G1 X306.853397 Y135.157047;
G1 X306.451592 Y135.050910;
G1 X306.047981 Y134.951869;
G1 X305.642687 Y134.859954;
G1 X305.235836 Y134.775192;
G1 X304.827556 Y134.697612;
G1 X304.417972 Y134.627236;
G1 X304.007212 Y134.564087;
G1 X303.595403 Y134.508184;
G1 X303.182673 Y134.459545;
G1 X302.769151 Y134.418185;
G1 X302.354963 Y134.384116;
G1 X301.940240 Y134.357349;
G1 X301.525110 Y134.337892;
G1 X301.109702 Y134.325752;
G1 X300.694144 Y134.320932;
G1 X300.628350 Y134.320840;
G1 X300.212780 Y134.324501;
G1 X299.797339 Y134.335482;
G1 X299.382157 Y134.353781;
G1 X298.967360 Y134.379391;
G1 X298.553080 Y134.412305;
G1 X298.139444 Y134.452512;
G1 X297.726580 Y134.500000;
G1 X297.314617 Y134.554754;
G1 X296.903682 Y134.616757;
G1 X296.493903 Y134.685990;
G1 X296.085408 Y134.762431;
G1 X295.678323 Y134.846057;
G1 X295.272774 Y134.936842;
G1 X294.868888 Y135.034757;
G1 X294.466789 Y135.139773;
G1 X294.066603 Y135.251856;
G1 X293.668453 Y135.370971;
G1 X293.272464 Y135.497083;
G1 X292.878757 Y135.630151;
G1 X292.487457 Y135.770134;
G1 X292.098683 Y135.916989;
G1 X291.712556 Y136.070671;
G1 X291.329197 Y136.231131;
G1 X290.948724 Y136.398320;
G1 X290.571256 Y136.572186;
G1 X290.196909 Y136.752675;
G1 X289.825800 Y136.939730;
G1 X289.458045 Y137.133295;
G1 X289.093756 Y137.333308;
G1 X288.733047 Y137.539708;
G1 X288.376031 Y137.752431;
G1 X288.022817 Y137.971410;
G1 X287.673516 Y138.196578;
G1 X287.328236 Y138.427864;
G1 X286.987085 Y138.665198;
G1 X286.650167 Y138.908505;
G1 X286.317588 Y139.157709;
G1 X285.989451 Y139.412734;
G1 X285.665858 Y139.673501;
G1 X285.346909 Y139.939927;
G1 X285.032703 Y140.211931;
G1 X284.723338 Y140.489429;
G1 X284.418909 Y140.772333;
G1 X284.119512 Y141.060557;
G1 X283.825239 Y141.354011;
G1 X283.536181 Y141.652603;
G1 X283.252429 Y141.956241;
G1 X282.974069 Y142.264831;
G1 X282.701190 Y142.578277;
G1 X282.433875 Y142.896482;
G1 X282.172207 Y143.219347;
G1 X281.916268 Y143.546771;
G1 X281.666136 Y143.878654;
G1 X281.421891 Y144.214891;
G1 X281.183607 Y144.555380;
G1 X280.951358 Y144.900013;
G1 X280.725217 Y145.248685;
G1 X280.505253 Y145.601286;
G1 X280.291536 Y145.957708;
G1 X280.084130 Y146.317839;
G1 X279.883102 Y146.681569;
G1 X279.688512 Y147.048783;
G1 X279.500422 Y147.419369;
G1 X279.318890 Y147.793211;
G1 X279.143972 Y148.170193;
G1 X278.975722 Y148.550198;
G1 X278.814194 Y148.933108;
G1 X278.659436 Y149.318804;
G1 X278.511497 Y149.707167;
G1 X278.370422 Y150.098076;
G1 X278.236257 Y150.491409;
G1 X278.109041 Y150.887045;
G1 X277.988816 Y151.284861;
G1 X277.875617 Y151.684733;
G1 X277.769480 Y152.086538;
G1 X277.670439 Y152.490149;
G1 X277.578524 Y152.895443;
G1 X277.493762 Y153.302294;
G1 X277.416182 Y153.710574;
G1 X277.345806 Y154.120158;
G1 X277.282657 Y154.530918;
G1 X277.226754 Y154.942727;
G1 X277.178115 Y155.355457;
G1 X277.136755 Y155.768979;
G1 X277.102686 Y156.183167;
G1 X277.075919 Y156.597890;
G1 X277.056462 Y157.013020;
G1 X277.044322 Y157.428428;
G1 X277.039502 Y157.843986;
G1 X277.039410 Y157.909780;
G1 X277.043071 Y158.325350;
G1 X277.054052 Y158.740791;
G1 X277.072351 Y159.155974;
G1 X277.097961 Y159.570770;
G1 X277.130875 Y159.985050;
G1 X277.171082 Y160.398686;
G1 X277.218570 Y160.811550;
G1 X277.273324 Y161.223514;
G1 X277.335327 Y161.634448;
G1 X277.404560 Y162.044227;
G1 X277.481001 Y162.452722;
G1 X277.564627 Y162.859807;
G1 X277.655412 Y163.265356;
G1 X277.753327 Y163.669242;
G1 X277.858343 Y164.071341;
G1 X277.970426 Y164.471527;
G1 X278.089541 Y164.869677;
G1 X278.215653 Y165.265666;
G1 X278.348721 Y165.659373;
G1 X278.488704 Y166.050673;
G1 X278.635559 Y166.439447;
G1 X278.789241 Y166.825574;
G1 X278.949701 Y167.208933;
G1 X279.116890 Y167.589406;
G1 X279.290756 Y167.966874;
G1 X279.471245 Y168.341221;
G1 X279.658300 Y168.712330;
G1 X279.851865 Y169.080086;
G1 X280.051878 Y169.444374;
G1 X280.258278 Y169.805083;
G1 X280.471001 Y170.162099;
G1 X280.689980 Y170.515313;
G1 X280.915148 Y170.864614;
G1 X281.146434 Y171.209894;
G1 X281.383768 Y171.551045;
G1 X281.627075 Y171.887963;
G1 X281.876279 Y172.220542;
G1 X282.131304 Y172.548679;
G1 X282.392071 Y172.872272;
G1 X282.658497 Y173.191221;
G1 X282.930501 Y173.505427;
G1 X283.207999 Y173.814792;
G1 X283.490903 Y174.119221;
G1 X283.779127 Y174.418618;
G1 X284.072581 Y174.712891;
G1 X284.371173 Y175.001949;
G1 X284.674811 Y175.285701;
G1 X284.983401 Y175.564061;
G1 X285.296847 Y175.836940;
G1 X285.615052 Y176.104255;
G1 X285.937917 Y176.365923;
G1 X286.265341 Y176.621862;
G1 X286.597224 Y176.871994;
G1 X286.933461 Y177.116239;
G1 X287.273950 Y177.354523;
G1 X287.618583 Y177.586772;
G1 X287.967255 Y177.812913;
G1 X288.319856 Y178.032877;
G1 X288.676278 Y178.246594;
G1 X289.036409 Y178.454000;
G1 X289.400139 Y178.655028;
G1 X289.767353 Y178.849618;
G1 X290.137939 Y179.037708;
G1 X290.511781 Y179.219240;
G1 X290.888763 Y179.394158;
G1 X291.268768 Y179.562408;
G1 X291.651678 Y179.723936;
G1 X292.037374 Y179.878694;
G1 X292.425737 Y180.026633;
G1 X292.816646 Y180.167708;
G1 X293.209979 Y180.301873;
G1 X293.605615 Y180.429089;
G1 X294.003431 Y180.549314;
G1 X294.403303 Y180.662513;
G1 X294.805108 Y180.768650;
G1 X295.208719 Y180.867691;
G1 X295.614013 Y180.959606;
G1 X296.020864 Y181.044368;
G1 X296.429144 Y181.121948;
G1 X296.838728 Y181.192324;
G1 X297.249488 Y181.255473;
G1 X297.661297 Y181.311376;
G1 X298.074027 Y181.360015;
G1 X298.487549 Y181.401375;
G1 X298.901737 Y181.435444;
G1 X299.316460 Y181.462211;
G1 X299.731590 Y181.481668;
G1 X300.146998 Y181.493808;
G1 X300.562556 Y181.498628;
G1 X300.628350 Y181.498720;
G1 X301.043920 Y181.495059;
G1 X301.459361 Y181.484078;
G1 X301.874543 Y181.465779;
G1 X302.289340 Y181.440169;
G1 X302.703620 Y181.407255;
G1 X303.117256 Y181.367048;
G1 X303.530120 Y181.319560;
G1 X303.942083 Y181.264806;
G1 X304.353018 Y181.202803;
G1 X304.762797 Y181.133570;
G1 X305.171292 Y181.057129;
G1 X305.578377 Y180.973503;
G1 X305.983926 Y180.882718;
G1 X306.387812 Y180.784803;
G1 X306.789911 Y180.679787;
G1 X307.190097 Y180.567704;
G1 X307.588247 Y180.448589;
G1 X307.984236 Y180.322477;
G1 X308.377943 Y180.189409;
G1 X308.769243 Y180.049426;
G1 X309.158017 Y179.902571;
G1 X309.544144 Y179.748889;
G1 X309.927503 Y179.588429;
G1 X310.307976 Y179.421240;
G1 X310.685444 Y179.247374;
G1 X311.059791 Y179.066885;
G1 X311.430900 Y178.879830;
G1 X311.798655 Y178.686265;
G1 X312.162944 Y178.486252;
G1 X312.523653 Y178.279852;
G1 X312.880669 Y178.067129;
G1 X313.233883 Y177.848150;
G1 X313.583184 Y177.622982;
G1 X313.928464 Y177.391696;
G1 X314.269615 Y177.154362;
G1 X314.606533 Y176.911055;
G1 X314.939112 Y176.661851;
G1 X315.267249 Y176.406826;
G1 X315.590842 Y176.146059;
G1 X315.909791 Y175.879633;
G1 X316.223997 Y175.607629;
G1 X316.533362 Y175.330131;
G1 X316.837791 Y175.047227;
G1 X317.137188 Y174.759003;
G1 X317.431461 Y174.465549;
G1 X317.720519 Y174.166957;
G1 X318.004271 Y173.863319;
G1 X318.282631 Y173.554729;
G1 X318.555510 Y173.241283;
G1 X318.822825 Y172.923078;
G1 X319.084493 Y172.600213;
G1 X319.340432 Y172.272789;
G1 X319.590564 Y171.940906;
G1 X319.834809 Y171.604669;
G1 X320.073093 Y171.264180;
G1 X320.305342 Y170.919547;
G1 X320.531483 Y170.570875;
G1 X320.751447 Y170.218274;
G1 X320.965164 Y169.861852;
G1 X321.172570 Y169.501721;
G1 X321.373598 Y169.137991;
G1 X321.568188 Y168.770777;
G1 X321.756278 Y168.400191;
G1 X321.937810 Y168.026349;
G1 X322.112728 Y167.649367;
G1 X322.280978 Y167.269362;
G1 X322.442506 Y166.886452;
G1 X322.597264 Y166.500756;
G1 X322.745203 Y166.112393;
G1 X322.886278 Y165.721484;
G1 X323.020443 Y165.328151;
G1 X323.147659 Y164.932515;
G1 X323.267884 Y164.534699;
G1 X323.381083 Y164.134827;
G1 X323.487220 Y163.733022;
G1 X323.586261 Y163.329411;
G1 X323.678176 Y162.924117;
G1 X323.762938 Y162.517266;
G1 X323.840518 Y162.108986;
G1 X323.910894 Y161.699402;
G1 X323.974043 Y161.288642;
G1 X324.029946 Y160.876833;
G1 X324.078585 Y160.464103;
G1 X324.119945 Y160.050581;
G1 X324.154014 Y159.636393;
G1 X324.180781 Y159.221670;
G1 X324.200238 Y158.806540;
G1 X324.212378 Y158.391132;
G1 X324.217198 Y157.975574;
G1 X324.217290 Y157.909780;
G1 X324.217290 Y157.909780;
Z30
G1 F1500.0 X0.000000 Y0.000000;

![IMG_1641](https://github.com/winder/Universal-G-Code-Sender/assets/155841882/a9f648ef-01d8-4220-a227-226c6954ef7a)

Thank you immensely for your help and valuable advice.

**Comment by breiler** [comment]
It looks like you have done everything you can do in software. So this might be a hardware problem. Make sure there is no play when moving the gantry (which could cause backlash) and go through and thighten any loose screws.

**Comment by Serhai** [comment]
Thank you for the answers. I will look into it .

**Comment by Serhai** [comment]
It's good, I did it! I hadn't properly adjusted the anti-backlash. I didn't know how it worked, and the machine's manual didn't explain it. It took me a month, plus assembling and disassembling my machine three times. I can't take it anymore! Thank you for your help!

**Reply by breiler** [reply]
I am glad you found it, let us hope it is downhill from here on! 😄
